src/server/util/lrucache.ts

```diff
@@ -0,0 +1,98 @@
+import _ from 'lodash'
+
+const MAX_SIZE_BYTES = 2000
+
+interface CookieData {
+  [shortUrl: string]: number
+}
+
+interface EpochToShortUrlMapping {
+  [epoch: number]: string
+}
+
+export default class LRUCache {
+  data: CookieData
+
+  lookupTable: EpochToShortUrlMapping = {}
+
+  timestamps: number[] = []
```

Comment:
checkout [this implementation](https://medium.com/sparkles-blog/a-simple-lru-cache-in-typescript-cba0d9807c40) of an LRUCache. it relies on the fact that Javascript `Map`s are actually ordered, so the implementation can be simplified significantly.

Reply:
actually the correct data structure here should be a Priority Queue.

when a redirect request arrives, the server simply iterates through the existing key-value pairs and insert all records, which will be automatically sorted, then upserts the existing short link/epoch (upsert is achieved by removing and reinserting).

to build up the new cookie, simply take from the PQ until iterator is exhausted or the size limit is hit.

does this make sense? if it does perhaps it will be easier to look for an existing implementation instead of rolling our own.

Reply:
Can look at Priority Queues from [google closure](https://google.github.io/closure-library/api/goog.structs.PriorityQueue.html) or [BucketJS](http://mauriciosantos.github.io/Buckets-JS/symbols/buckets.PriorityQueue.html)

Reply:
thinking yet again - maybe we're over-engineering here. looking at this pattern:

```javascript
const cache = new LRUCache(req.session!.visits)
cache.updateEntry(shortUrl)
req.session!.visits = cache.getData()
```

we simply need a reducer function

```javascript
req.session!.visits = updateCookies(req.session!.visits || {}, shortUrl)
```

where

```javascript
type updateCookies = (visits: Record<string, number>, shortUrl: string) => Record<string, number>
```